src/datasets/utils/tf_utils.py

```diff
@@ -68,3 +80,423 @@ def is_numeric_feature(feature):
+            for col, array in shape_arrays.items():
+                array[:] = -1
+            array_ready_event.set()
+
+    @staticmethod
+    def distribute_batches(dataset, batch_size, drop_remainder, num_workers, shuffle):
+        indices = np.arange(len(dataset))
+        if shuffle:
+            np.random.shuffle(indices)
+        num_samples = len(indices)
+        # We distribute the batches so that reading from the workers in round-robin order yields the exact
+        # order specified in indices. This is only important when shuffle is False, but we do it regardless.
+        incomplete_batch_cutoff = num_samples - (num_samples % batch_size)
+        indices, last_incomplete_batch = np.split(indices, [incomplete_batch_cutoff])
+        if drop_remainder or len(last_incomplete_batch) == 0:
+            last_incomplete_batch = None
+
+        indices = indices.reshape(-1, batch_size)
+        num_batches = len(indices)
+        final_batches_cutoff = num_batches - (num_batches % num_workers)
+        indices, final_batches = np.split(indices, [final_batches_cutoff])
+        indices = indices.reshape(-1, num_workers, batch_size)
+
+        per_worker_indices = np.split(indices, indices.shape[1], axis=1)
+        per_worker_indices = [np.squeeze(worker_indices, 1) for worker_indices in per_worker_indices]
+        # Distribute the final batches to the first workers
+        for i in range(len(final_batches)):
+            # len(final_batches) can be zero, and is always less than num_workers
+            per_worker_indices[i] = np.concatenate([per_worker_indices[i], final_batches[i].reshape(1, -1)], axis=0)
+        # Add the last incomplete batch to the next worker, which might be the first worker
+        if last_incomplete_batch is not None:
+            incomplete_batch_worker_idx = len(final_batches)
+        else:
+            incomplete_batch_worker_idx = None
+        return per_worker_indices, last_incomplete_batch, incomplete_batch_worker_idx
+
+
+def multiprocess_dataset_to_tf(
+    dataset,
+    cols_to_retain,
+    collate_fn,
+    collate_fn_args,
+    columns_to_np_types,
+    output_signature,
+    shuffle,
+    batch_size,
+    drop_remainder,
+    num_workers,
+):
+    if config.TF_AVAILABLE:
+        import tensorflow as tf
+    else:
+        raise ImportError("Called a Tensorflow-specific function but Tensorflow is not installed.")
+
+    data_generator = NumpyMultiprocessingGenerator(
+        dataset=dataset,
+        cols_to_retain=cols_to_retain,
+        collate_fn=collate_fn,
+        collate_fn_args=collate_fn_args,
+        columns_to_np_types=columns_to_np_types,
+        output_signature=output_signature,
+        shuffle=shuffle,
+        batch_size=batch_size,
+        drop_remainder=drop_remainder,
+        num_workers=num_workers,
+    )
+
+    tf_dataset = tf.data.Dataset.from_generator(data_generator, output_signature=output_signature)
```

Comment:
Does using from_generator have consequences on how the dataset can be used compared to from_tensor_slices ?
If so, this must be documented

Reply:
No, from a UX perspective the output dataset should be identical!